File: nodejs-tools/nodejs-compiler-tools/formatter-xml/src/format_xml_prolog_node.ts
import { XmlPrologNode } from '@vighnesh153/parser-xml';
import { buildIndentationSpace } from './build_indentation_space.ts';
import { formatXmlElementAttribute } from './format_xml_element_attribute.ts';

type FormatXmlPrologNodeOptions = {
  xmlPrologNode: XmlPrologNode;
  indentationLevel: number;
  indentation: number;
};

export function formatXmlPrologNode({
  xmlPrologNode,
  indentationLevel,
  indentation,
}: FormatXmlPrologNodeOptions): string {
  const localStringBuilder: string[] = [];

  localStringBuilder.push(buildIndentationSpace({ indentationLevel, indentation }) + `<?xml`);

  for (const attribute of xmlPrologNode.attributes) {
    localStringBuilder.push(formatXmlElementAttribute(attribute));
  }

  return localStringBuilder.join(' ') + `?>`;
}

File: tools-jvm/XmlFormatter/src/main/kotlin/dev/vighnesh153/xml/formatter/fmt/FormatXmlPrologNode.kt
package dev.vighnesh153.xml.formatter.fmt

import dev.vighnesh153.xml.formatter.ast.XmlPrologNode
import dev.vighnesh153.xml.formatter.ast.buildIndentation

internal fun XmlPrologNode.format(
    indentationLevel: Int,
    indentation: Int,
): String {
    val stringBuilder = StringBuilder()

    stringBuilder.append(
        buildIndentation(
            indentationLevel = indentationLevel,
            indentation = indentation
        )
    )
    stringBuilder.append("<?xml")

    for (attr in attributes) {
        stringBuilder.append(" ${attr.format()}")
    }

    return "$stringBuilder?>"
}